### PR TITLE
PIM-7968: Fix users

### PR DESCRIPTION
**Pull request by anaelChardan**
**Description (for Contributor and Core Developer)**

This PR makes retrocompatibility with 2.3 and fixes some glitches.

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | -
| Added legacy Behats               | -
| Added acceptance tests            | -
| Added integration tests           | -
| Changelog updated                 | -
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | OK
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/Platform/Bundle/UIBundle/Resources/public/js/form/common/save-form.js [hunk 1]
```diff
@@ -89,18 +91,22 @@ define(
                 this.showLoadingMask();
                 this.getRoot().trigger('pim_enrich:form:entity:pre_save');
 
+                let entityIdProperty = this.config.entityIdentifierParamName || 'code';
```
Comment (damien-carcel) [inline src/Akeneo/Platform/Bundle/UIBundle/Resources/public/js/form/common/save-form.js]: those 3 should be `const`, not `let`

File: src/Akeneo/UserManagement/Bundle/Controller/Rest/UserController.php [hunk 2]
```diff
@@ -263,19 +275,39 @@ public function deleteAction(Request $request, $identifier): Response
         return new Response(null, Response::HTTP_NO_CONTENT);
     }
 
-    private function getUserOr404($username): ?UserInterface
+    private function getUserOr404($identifier): ?UserInterface
```
Comment (momoss) [inline src/Akeneo/UserManagement/Bundle/Controller/Rest/UserController.php]: just return UserInterface ;-)

File: src/Akeneo/Platform/Bundle/UIBundle/Resources/public/js/saver/entity-saver.js [hunk 3]
```diff
@@ -32,6 +36,17 @@ define([
             setUrl: function (url) {
                 this.url = url;
 
+                return this;
+            },
+
+            /**
+             * Sets the identifierProperty for the url
+             *
+             * @param {string }identifierProperty
```
Comment (pierallard) [inline src/Akeneo/Platform/Bundle/UIBundle/Resources/public/js/saver/entity-saver.js]: error space

File: src/Akeneo/UserManagement/Bundle/Controller/Rest/UserController.php [hunk 4]
```diff
@@ -263,19 +275,39 @@ public function deleteAction(Request $request, $identifier): Response
...
+    private function validatePasswordCreate(array $data): ConstraintViolationListInterface
+    {
+        $violations = [];
+
+        if (!isset($data['password'])) {
+            return new ConstraintViolationList([]);
+        }
+
+        if (($data['password_repeat'] ?? '') !== $data['password']) {
+            $violations[] = new ConstraintViolation('Password does not match', '', [], '', 'password_repeat', '');
```
Comment (pierallard) [inline src/Akeneo/UserManagement/Bundle/Controller/Rest/UserController.php]: "Passwords"

File: src/Akeneo/UserManagement/Bundle/Resources/public/js/form/delete.js [hunk 5]
```diff
@@ -9,7 +9,11 @@
  */
 define(['pim/form/common/delete', 'pim/remover/user'], function (DeleteForm, UserRemover) {
     return DeleteForm.extend({
-        remover: UserRemover
+        remover: UserRemover,
+
+        getIdentifier: function () {
```
Comment (pierallard) [inline src/Akeneo/UserManagement/Bundle/Resources/public/js/form/delete.js]: doc

File: tests/legacy/features/Behat/Context/NavigationContext.php [hunk 6]
```diff
@@ -275,7 +276,24 @@ public function iAmOnTheEntityEditPage($identifier, $page)
      * @param string $page
      *
      * @Given /^I edit the "([^"]*)" (user)$/
-     * @Given /^I am on the "([^"]*)" (channel|family|user) page$/
+     * @Given /^I am on the "([^"]*)" (user) page$/
+     */
+    public function iAmOnTheUserEditPage($identifier, $page)
+    {
+        /** @var User $user */
```
Comment (pierallard) [inline tests/legacy/features/Behat/Context/NavigationContext.php]: remove

File: src/Akeneo/UserManagement/Bundle/Controller/Rest/UserController.php [hunk 7]
```diff
@@ -263,19 +275,39 @@ public function deleteAction(Request $request, $identifier): Response
...
 
         return $user;
     }
 
+    /**
+     * @param array $data
+     *
+     * @return ConstraintViolationListInterface
+     */
+    private function validatePasswordCreate(array $data): ConstraintViolationListInterface
```
Comment (skeleton) [inline src/Akeneo/UserManagement/Bundle/Controller/Rest/UserController.php]: I don't understand why this is done in this PR. Was it something missing previously? There is no symfony validator that can doing this stuff?

File: src/Akeneo/UserManagement/Component/Updater/UserUpdater.php [hunk 8]
```diff
@@ -437,7 +436,7 @@ private function setAvatar($user, $data)
                     $data['filePath']
                 );
             }
-            $fileInfo = $this->fileStorer->store($rawFile, FileStorage::CATALOG_STORAGE_ALIAS);
+            $fileInfo = $this->fileStorer->store($rawFile, 'catalogStorage');
```
Comment (skeleton) [inline src/Akeneo/UserManagement/Component/Updater/UserUpdater.php]: why do you hardcode it?

File: src/Akeneo/Platform/Bundle/UIBundle/Resources/public/js/form/common/save-form.js (hunk 1 above)
Comment (anaelChardan) [reply src/Akeneo/Platform/Bundle/UIBundle/Resources/public/js/form/common/save-form.js]: done ;)

File: src/Akeneo/UserManagement/Bundle/Controller/Rest/UserController.php (hunk 2 above)
Comment (anaelChardan) [reply src/Akeneo/UserManagement/Bundle/Controller/Rest/UserController.php]: done

File: CHANGELOG-3.0.md [hunk 9]
```diff
@@ -16,6 +16,7 @@
 - PIM-7839: Remove date of birth
 - GITHUB-8234 & GITHUB-8383: Fix constraints on attribute code. Cheers @oliverde8 & @navneetbhardwaj !
 - TIP-1018: Adds a script to check container services instantiability (bin/check-services-instantiability)
+- PIM-7968: Make the user fetchable by id instead of username
```
[the remark this answers is not in the document]
Comment (anaelChardan) [reply CHANGELOG-3.0.md]: not everywhere, it seems that in the twa it was doing some stuff weirdly ^^

File: src/Akeneo/UserManagement/Bundle/Controller/Rest/UserController.php (hunk 7 above)
Comment (anaelChardan) [reply src/Akeneo/UserManagement/Bundle/Controller/Rest/UserController.php]: Yeah it was missing and due to internal manipulation we cannot use the validator.

File: src/Akeneo/UserManagement/Component/Updater/UserUpdater.php (hunk 8 above)
Comment (anaelChardan) [reply src/Akeneo/UserManagement/Component/Updater/UserUpdater.php]: to avoid the coupling with enrichment, actually it didnt work at all before.

**Pull request merged**